onyo/lib/onyo.py

```diff
@@ -1240,3 +1241,128 @@ def unset(self, paths: Iterable[Union[Path, str]], values: str,
             self.restore()
 
         return diff
+
+    #
+    # GET
+    #
+    # TODO: This should be a staticmethod in utils
+    def _walk(self, path, depth) -> Generator:
+        """Recursively yield filepaths up to a given depth"""
+        depth -= 1
+        with os.scandir(path) as p:
+            for entry in p:
+                # ignore hidden files and folders
+                # TODO: Better to use _is_protected_path or ignore hidden?
+                #   `if self._is_protected_path(entry.path):`
+                if entry.name.startswith('.'):
+                    continue
+
+                # yield discovered assets
+                if not entry.is_dir():
+                    yield entry.path
+
+                # explore dirs if depth allows
+                if entry.is_dir() and depth >= 0:
+                    yield from self._walk(entry.path, depth)
+
+    @staticmethod
+    def _filter_by_name(assets: List[str], filters: dict) -> Generator:
+        """Filter assets by filename to prevent redundant asset loading"""
+        keys = ['type', 'make', 'model', 'serial']
+        for asset in assets:
+            split = set(zip(keys, re.split('[_.]', os.path.basename(asset))))
+            if set(filters.items()).issubset(split):
+                yield asset
+
+    @staticmethod
+    def _get_asset_data(
+            assets: Union[Generator[Any, Any, Any], List[Any]], keys: Set[str],
+            filters: Union[dict, None] = None) -> Generator:
+        """Return data from asset keys for assets matching filters"""
+        yaml = YAML(typ='safe')
+        default_keys = ['type', 'make', 'model', 'serial']
+        string_types = {  # TODO: Do we have a list of accepted string types?
+            '[int]': int, '[float]': float, '[str]': str, '[list]': list,
+            '[dict]': dict, '[bool]': bool}
+
+        for asset in assets:
+            try:
+                data = yaml.load(Path(asset))
+            except scanner.ScannerError as e:
+                log.error(e)
+                continue
+
+            # add type,make,model,serial to data if key does not exist
```

Comment:
The "pseudo keys" should not also exist in the asset. If they do, they should be ignored. (This should be an `fsck` check. I opened #283).

Reply:
I've changed it so that the key values in the asset name will overwrite any existing in the asset data.